test/integration/wan_lan_dht_test.go

```diff
@@ -0,0 +1,277 @@
+	if err != nil {
+		return err
+	}
+	defer testPeer.Close()
+
+	wanPeers := []*core.IpfsNode{}
+	lanPeers := []*core.IpfsNode{}
+
+	connectionContext, connCtxCancel := context.WithTimeout(ctx, 15*time.Second)
+	defer connCtxCancel()
+	for i := 0; i < numPeers; i++ {
+		wanPeer, err := core.NewNode(ctx, &core.BuildCfg{
+			Online:  true,
+			Routing: libp2p2.DHTServerOption,
+			Host:    mock.MockHostOption(mn),
+		})
+		if err != nil {
+			return err
+		}
+		defer wanPeer.Close()
+		wanAddr := makeAddr(uint32(i), true)
+		wanPeer.Peerstore.AddAddr(wanPeer.Identity, wanAddr, peerstore.PermanentAddrTTL)
+		for _, p := range wanPeers {
+			_, _ = mn.LinkPeers(p.Identity, wanPeer.Identity)
+			_ = wanPeer.PeerHost.Connect(connectionContext, p.Peerstore.PeerInfo(p.Identity))
+		}
+		wanPeers = append(wanPeers, wanPeer)
+
+		lanPeer, err := core.NewNode(ctx, &core.BuildCfg{
+			Online: true,
+			Host:   mock.MockHostOption(mn),
+		})
+		if err != nil {
+			return err
+		}
+		defer lanPeer.Close()
+		lanAddr := makeAddr(uint32(i), false)
+		lanPeer.Peerstore.AddAddr(lanPeer.Identity, lanAddr, peerstore.PermanentAddrTTL)
+		for _, p := range lanPeers {
+			_, _ = mn.LinkPeers(p.Identity, lanPeer.Identity)
+			_ = lanPeer.PeerHost.Connect(connectionContext, p.Peerstore.PeerInfo(p.Identity))
+		}
+		lanPeers = append(lanPeers, lanPeer)
+	}
+	connCtxCancel()
+
+	// Add interfaces / addresses to test peer.
+	wanAddr := makeAddr(0, true)
+	testPeer.Peerstore.AddAddr(testPeer.Identity, wanAddr, peerstore.PermanentAddrTTL)
+	lanAddr := makeAddr(0, false)
+	testPeer.Peerstore.AddAddr(testPeer.Identity, lanAddr, peerstore.PermanentAddrTTL)
+
+	// The test peer is connected to one lan peer.
+	for _, p := range lanPeers {
+		if _, err := mn.LinkPeers(testPeer.Identity, p.Identity); err != nil {
+			return err
+		}
+	}
+	err = testPeer.PeerHost.Connect(ctx, lanPeers[0].Peerstore.PeerInfo(lanPeers[0].Identity))
+	if err != nil {
+		return err
+	}
+
+	startupCtx, startupCancel := context.WithTimeout(ctx, time.Second*60)
+StartupWait:
+	for {
+		select {
+		case err := <-testPeer.DHT.LAN.RefreshRoutingTable():
+			if err != nil {
+				fmt.Printf("Error refreshing routing table: %v\n", err)
+			}
+			if testPeer.DHT.LAN.RoutingTable() == nil ||
+				testPeer.DHT.LAN.RoutingTable().Size() == 0 ||
+				err != nil {
+				time.Sleep(100 * time.Millisecond)
+				continue
+			}
+			break StartupWait
+		case <-startupCtx.Done():
+			startupCancel()
+			return fmt.Errorf("expected faster dht bootstrap")
+		}
+	}
+	startupCancel()
+
+	// choose a lan peer and validate lan DHT is functioning.
+	i := rand.Intn(len(lanPeers))
+	if testPeer.PeerHost.Network().Connectedness(lanPeers[i].Identity) == corenet.Connected {
+		i = (i + 1) % len(lanPeers)
+		if testPeer.PeerHost.Network().Connectedness(lanPeers[i].Identity) == corenet.Connected {
+			_ = testPeer.PeerHost.Network().ClosePeer(lanPeers[i].Identity)
+			testPeer.PeerHost.Peerstore().ClearAddrs(lanPeers[i].Identity)
+		}
+	}
```

Comment:
Now that we're trying to find a record, not a peer, this shouldn't be necessary.